File: project2/Plant.java
import java.util.Random;

abstract class Plant extends Organism {
    private static final long serialVersionUID = 1L;
    private static final Random random = new Random();
    private static final int WIDTH = 40;
    private static final int HEIGHT = 20;
    
    public Plant(int str, int x, int y, World w) {
        super(str, 0, x, y, w);
    }
    
    @Override
    public void action() {
        if (random.nextInt(10) == 0) {
            int nx = x + (random.nextInt(3) - 1);
            int ny = y + (random.nextInt(3) - 1);
            if (nx >= 0 && ny >= 0 && nx < WIDTH && ny < HEIGHT && world.grid[ny][nx] == null) {
                try {
                    world.queueOrganism(this.getClass().getDeclaredConstructor(int.class, int.class, World.class)
                            .newInstance(nx, ny, world));
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
        age++;
    }
    
    @Override
    public void collision(Organism other) {
        alive = false;
    }
    
    @Override
    public char draw() {
        return 'P';
    }
}

File: project3/plant.py
from organism import Organism
import random

class Plant(Organism):
    WIDTH = 40
    HEIGHT = 20
    
    def __init__(self, strength, x, y, world):
        super().__init__(strength, 0, x, y, world)
    
    def action(self):
        # Plants have a 10% chance to spread to adjacent cells
        if random.randint(0, 9) == 0:
            nx = self.x + (random.randint(0, 2) - 1)
            ny = self.y + (random.randint(0, 2) - 1)
            
            if (0 <= nx < self.WIDTH and 0 <= ny < self.HEIGHT and 
                self.world.grid[ny][nx] is None):
                # Create a new plant of the same type
                new_plant = self.__class__(nx, ny, self.world)
                self.world.queue_organism(new_plant)
        
        self.age += 1
    
    def collision(self, other):
        self.alive = False
    
    def draw(self):
        return 'P'
